apps/prairielearn/src/tests/render.test.ts

```diff
@@ -0,0 +1,365 @@
+
+import { assert } from 'chai';
+import fs from 'fs-extra';
+import { HTMLRewriter } from 'html-rewriter-wasm';
+import { HtmlValidate } from 'html-validate';
+import { HTMLHint } from 'htmlhint';
+
+import { config } from '../lib/config.js';
+import type { Course, Question, Submission, Variant } from '../lib/db-types.js';
+import { features } from '../lib/features/index.js';
+import { buildQuestionUrls } from '../lib/question-render.js';
+import { makeVariant } from '../lib/question-variant.js';
+import * as questionServers from '../question-servers/index.js';
+import * as helperServer from '../tests/helperServer.js';
+
+const htmlvalidate = new HtmlValidate();
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+
+const exampleCoursePath = resolve(__dirname, '..', '..', '..', '..', 'exampleCourse');
+const questionsPath = join(exampleCoursePath, 'questions');
+
+// Helper function to find question directories recursively
+const findQuestionDirectories = (dir: string): string[] => {
+  const questionDirs: string[] = [];
+  const entries = fs.readdirSync(dir, { withFileTypes: true, recursive: true });
+  for (const entry of entries) {
+    if (entry.isDirectory()) {
+      if (fs.existsSync(join(entry.parentPath, entry.name, 'info.json'))) {
+        questionDirs.push(join(entry.parentPath, entry.name));
+      }
+    }
+  }
+  return questionDirs;
+};
+
+const rewriteValidatorFalsePositives = async (html: string): Promise<string> => {
+  /**
+   * Bootstrap uses JS to set accessibility attributes like aria-label on elements.
+   * We rewrite the HTML to avoid incorrectly flagged errors in HTMLValidate.
+   *
+   * Additionally, pl-drawing uses JS to set the img src attribute.
+   *
+   * pl-code has span elements with empty style attributes we need to ignore.
+   * pl-overlay has empty style attributes we need to ignore.
+   */
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+  let output = '';
+  const rewriter = new HTMLRewriter((chunk) => {
+    output += decoder.decode(chunk);
+  });
+  rewriter
+    .on('a, button', {
+      element(el) {
+        if (el.hasAttribute('aria-label')) return;
+        const title = el.getAttribute('data-bs-title') ?? el.getAttribute('title');
+        if (title) {
+          el.setAttribute('aria-label', title);
+        }
+      },
+    })
+    .on('.pl-drawing-button img', {
+      element(el) {
+        const src = el.getAttribute('src');
+        if (src === '') {
+          el.setAttribute('src', 'pl-drawing-dummy-img.png');
+        }
+      },
+    })
+    .on('.linenodiv pre span, .pl-overlay', {
+      element(el) {
+        const style = el.getAttribute('style');
+        if (style?.trim() === '') {
+          el.removeAttribute('style');
+        }
+      },
+    });
+  await rewriter.write(encoder.encode(html));
+  await rewriter.end();
+  return output;
+};
+
+const validateHtml = async (html: string) => {
```

Comment:
In `apps/prairielearn/src/tests/accessibility/index.test.ts`, we also run some automated accessibility tests. Are you thinking at all about doing the same here?

Reply:
Good idea, I'll add the `axe` checks here as well.

Reply:
Now added.